plugin-c/crypto.cpp

```diff
@@ -0,0 +1,253 @@
+using namespace libff;
+
+extern "C" {
+struct blockheader {
+  uint64_t len;
+};
+
+struct string {
+  struct blockheader b;
+  unsigned char data[0];
+};
+
+static string *hexEncode(unsigned char *digest, size_t len) {
+  uint64_t hexLen = len * 2;
+  char byte[3];
+  struct string *result = (struct string *)koreAllocToken(hexLen+8);
+  result->b.len = hexLen;
+  for (size_t i = 0, j = 0; i < len; i++, j += 2) {
+    sprintf(byte, "%02x", digest[i]);
+    result->data[j] = byte[0];
+    result->data[j+1] = byte[1];
+  }
+  return result;
+}
+
+struct string *hook_KRYPTO_keccak256(struct string *str) {
+  Keccak_256 h;
+  unsigned char digest[32];
+  h.CalculateDigest(digest, str->data, len(str));
+  return hexEncode(digest, sizeof(digest));
+}
+
+struct string *hook_KRYPTO_sha256(struct string *str) {
+  SHA256 h;
+  unsigned char digest[32];
+  h.CalculateDigest(digest, str->data, len(str));
+  return hexEncode(digest, sizeof(digest));
+}
+
+struct string *hook_KRYPTO_ripemd160(struct string *str) {
+  RIPEMD160 h;
+  unsigned char digest[20];
+  h.CalculateDigest(digest, str->data, len(str));
+  return hexEncode(digest, sizeof(digest));
+}
+
+struct string *hook_KRYPTO_ecdsaRecover(struct string *str, mpz_t v, struct string *r, struct string *s) {
+  if (len(str) != 32 || len(r) != 32 || len(s) != 32) {
+    return hexEncode(nullptr, 0);
+  }
+  unsigned char sigArr[64];
+  memcpy(sigArr, r->data, 32);
+  memcpy(sigArr+32, s->data, 32);
+  secp256k1_context *ctx = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY | SECP256K1_CONTEXT_SIGN);
+  if (!mpz_fits_ulong_p(v)) {
+    return hexEncode(nullptr, 0);
+  }
+  unsigned long v_long = mpz_get_ui(v);
+  if (v_long < 27 || v_long > 28) {
+    return hexEncode(nullptr, 0);
+  }
+  secp256k1_ecdsa_recoverable_signature sig;
+  if (!secp256k1_ecdsa_recoverable_signature_parse_compact(ctx, &sig, sigArr, v_long - 27)) {
+    return hexEncode(nullptr, 0);
+  }
+  secp256k1_pubkey key;
+  if (!secp256k1_ecdsa_recover(ctx, &key, &sig, str->data)) {
+    return hexEncode(nullptr, 0);
+  }
+  unsigned char serialized[65];
+  size_t len = sizeof(serialized);
+  secp256k1_ec_pubkey_serialize(ctx, serialized, &len, &key, SECP256K1_EC_UNCOMPRESSED);
+  struct string *result = (struct string *)koreAlloc(sizeof(struct string) + 64);
+  result->b.len = 64;
+  memcpy(result->data, serialized+1, 64);
+  return result;
+}
+
+struct g1point {
+  struct blockheader h;
+  mpz_ptr x;
+  mpz_ptr y;
+};
+
+struct g2point {
+  struct blockheader h;
+  mpz_ptr x0;
+  mpz_ptr x1;
+  mpz_ptr y0;
+  mpz_ptr y1;
+};
+
+struct inj {
+  struct blockheader h;
+  void *data;
+};
+
+struct kseq {
+  struct blockheader h;
+  inj *inj;
+  void *dotk;
+};
+
+bool bn128_initialized = false;
+
+extern "C++" {
+
+void initBN128() {
+  if (bn128_initialized) {
+    return;
+  }
+  alt_bn128_pp::init_public_params();
+  inhibit_profiling_info = true;
+  bn128_initialized = true;
+}
+
+alt_bn128_G1 getPoint(g1point *pt) {
```

Comment:
I don't think this function is intended for external use. Should we reflect that somehow? Maybe `-fvisibility=hidden` and `__attribute__ ((visibility ("default")))`, maybe just `static` functions?